src/textual/widgets/text_input.py

```diff
@@ -0,0 +1,172 @@
+        elif key == "home":
+            self.cursor_index = 0
+        elif key == "end":
+            self.cursor_index = len(self.current_text)
+        elif key not in Keys.values():
+            new_text = (
+                self.current_text[: self.cursor_index]
+                + key
+                + self.current_text[self.cursor_index :]
+            )
+            self.current_text = new_text
+            self.cursor_index = min(len(self.current_text), self.cursor_index + 1)
+            changed = True
+
+        if changed:
+            self.post_message_no_wait(self.Changed(self, value=self.current_text))
+
+
+class TextInput(TextInputBase, can_focus=True):
+    CSS = """
+    TextInput {
+        width: auto;
+        background: $primary;
+        height: 3;
+        padding: 0 1;
+        content-align: left middle;
+    }
+
+    TextInput:hover {
+        background: $primary-darken-1;
+    }
+
+    TextInput:focus {
+        background: $primary-darken-2;
+        border: solid $primary-lighten-1;
+        padding: 0;
+    }
+
+    App.-show-focus TextInput:focus {
+        tint: $accent 20%;
+    }
+    """
+
+    class Submitted(Message, bubble=True):
+        def __init__(self, sender: MessageTarget, value: str) -> None:
+            """Message posted when the user presses the 'enter' key while
+            focused on a TextInput widget.
+
+            Args:
+                sender (MessageTarget): Sender of the message
+                value (str): The value in the TextInput
+            """
+            super().__init__(sender)
+            self.value = value
+
+    def __init__(
+        self,
+        placeholder: str = "",
+        initial: str = "",
+        *,
+        name: str | None = None,
+        id: str | None = None,
+        classes: str | None = None,
+    ):
+        super().__init__(name=name, id=id, classes=classes)
+        self.placeholder = placeholder
+        self.current_text = initial if initial else ""
+
+    def get_content_width(self, container_size: Size, viewport_size: Size) -> int:
+        return (
+            max(len(self.current_text), len(self.placeholder))
+            + self.styles.gutter.width
+        )
+
+    def render(self) -> RenderableType:
+        # We only show the cursor if the widget has focus
+        show_cursor = self.has_focus
+        if self.current_text:
+            display_text = Text(self.current_text, no_wrap=True)
+
+            if show_cursor:
+                display_text = self._apply_cursor_to_text(
+                    display_text, self.cursor_index
+                )
+            return display_text
+        else:
+            # The user has not entered text - show the placeholder
+            display_text = Text(self.placeholder, "dim")
+            if show_cursor:
+                display_text = self._apply_cursor_to_text(display_text, 0)
+            return display_text
+
+    def _apply_cursor_to_text(self, display_text: Text, index: int):
```

Comment:
Cursor implementation applies `reverse` style to the cursor location. If the cursor is at the very end (character append position), then we use the Unicode block character to represent it.